pennylane/hf/tapering.py

```diff
@@ -0,0 +1,285 @@
+
+        # swap rows krow and irow
+        rref_binary_matrix[krow], rref_binary_matrix[irow] = deepcopy(
+            rref_binary_matrix[irow]
+        ), deepcopy(rref_binary_matrix[krow])
+
+        # store remainder columns of the row irow
+        pvtcols = rref_binary_matrix[irow, icol:]
+
+        # get the column icol and set its irow element to 0 to avoid XORing pivot row with itself
+        currcol = deepcopy(rref_binary_matrix[:, icol])
+        currcol[irow] = 0
+        rref_binary_matrix[:, icol:] ^= np.outer(currcol, pvtcols)
+
+    return rref_binary_matrix.astype(int)
+
+
+def _kernel(binary_matrix):
+    r"""Computes the kernel of a binary matrix on the binary finite field :math:`\mathbb{Z}_2`.
+
+    Args:
+        binary_matrix (array[int]): binary matrix representation of the Hamiltonian
+    Returns:
+        array[int]: nullspace of the `binary_matrix` where each row correspond to a
+        basis vector in the nullspace
+
+    **Example**
+
+    .. code-block:: python
+
+        >>> binary_matrix = np.array([[1, 0, 0, 0, 0, 1, 0, 0],
+                                      [0, 0, 1, 1, 1, 1, 1, 1],
+                                      [0, 0, 0, 1, 1, 0, 0, 1]])
+        >>> _kernel(binary_matrix)
+         array([[0, 1, 0, 0, 0, 0, 0, 0],
+                [0, 0, 1, 1, 1, 0, 0, 0],
+                [1, 0, 1, 0, 0, 1, 0, 0],
+                [0, 0, 1, 0, 0, 0, 1, 0],
+                [0, 0, 1, 1, 0, 0, 0, 1]])
+
+    """
+
+    # Get the columns with and without pivots
+    pivots = (binary_matrix.T != 0).argmax(axis=0)
+    nonpivots = np.setdiff1d(range(len(binary_matrix[0])), pivots)
+
+    # Initialize the nullspace
+    null_vector = np.zeros((binary_matrix.shape[1], len(nonpivots)), dtype=int)
+    null_vector[nonpivots, np.arange(len(nonpivots))] = 1
+
+    # Fill up the nullspace vectors from the binary matrix
+    null_vector_indices = np.ix_(pivots, np.arange(len(nonpivots)))
+    binary_vector_indices = np.ix_(np.arange(len(pivots)), nonpivots)
+    null_vector[null_vector_indices] = -binary_matrix[binary_vector_indices] % 2
+
+    nullspace = null_vector.T
+    return nullspace
+
+
+def get_generators(nullspace, num_qubits):
+    r"""Compute the generators :math:`\{\tau_1, \ldots, \tau_k\}` from the nullspace of
+    the binary matrix form of a Hamiltonian over the binary field :math:`\mathbb{Z}_2`.
+    These correspond to the generator set of the :math:`\mathbb{Z}_2`-symmetries present
+    in the Hamiltonian as given in `arXiv:1910.14644 <https://arxiv.org/abs/1910.14644>`_.
+
+    Args:
+        nullspace (array[int]): kernel of the binary matrix corresponding to the Hamiltonian
+        num_qubits (int): number of wires required to define the Hamiltonian
+
+    Returns:
+        list[pennylane.Hamiltonian]: list of generators of symmetries, taus, for the Hamiltonian
+
+    **Example**
+
+    .. code-block:: python
+
+        >>> kernel = np.array([[0, 1, 0, 0, 0, 0, 0, 0],
```

Comment:
In the documentation, fro some reason the first row of this array appears in a different colour than the rest :thinking: Not sure why, but we should fix it

Reply:
Fixed. I had forgotten the `...` while writing the array in a multiline fashion.